[RFC] thinLTO for SYCL

sarnex opened the pull request:
Hi all,

I'd like to get some architecture and code-organzation level comments on my SYCL thinLTO prototype.

The feature isn't ready for code-review, so I'd appreciate keeping the discussion somewhat high level.

Firstly, if someone wants to try this, you can just clone my branch, build the compiler with `--cmake-opt="-DLLVM_EXPERIMENTAL_TARGETS_TO_BUILD=SPIRV"`, and then compile SYCL programs with `--offload-new-driver -foffload-lto=thin`. 

The main idea of the change is as follows:

We hook into the existing LTO framework in `clang-linker-wrapper` by adding a callback that runs right before the backend runs (so right before the SPIR-V backend would be called).

Before the callback is run, function importing happens automatically inside the thinLTO framework where we link in the definitions for any functions referenced in this module but defined in another. When the callback is run, the module passed in should have everything linked in and be ready for processing.

In that callback, we do all the required processing of the SYCL module. In the current prototype, that includes:
Generating the module properties
Generating the module symbol table
Calling the SPIR-V translator

In the future, we will also need to handle at least spec constants and internalization of non-entry points.

The final step of the callback, calling the SPIR-V translator, should be removed when the SPIR-V backend is ready. 

One problem is that we need to store the module properties, symbol table, and SPIR-V translator output file path until a later time where we are ready to create the final images, which requires all of this information.

We are also somewhat limited on what we can do to store the data, because the callbacks run on threads inside the LTO framework, and some data structures (like `StringSaver`) are not thread-safe. We need to do this information inside the LTO framework thread because that is the only point we have the IR with functions imported.

As a low-effort hack to get the prototype working, I'm just storing that information as `std::string`, which is pretty inefficient and gross.

I decided to store the information inside the `OffloadBinary`. Otherwise, we'd need to create and pass a SYCL-specific data structure all over the place, which seems bad. The current changes to `OffloadBinary` to add a temporary string vector is definitely disgusting, but I'm proposing we make some change to the class to allow us to store the information we need, but I don't have a good idea for a clean and thread-safe data structure to use. I don't have a better idea for a place to store all this information.

In terms of test results:

Here are the results when running E2E tests on PVC with `--offload-new-driver` only enabled:
```
Total Discovered Tests: 2220
  Unsupported      :  543 (24.46%)
  Passed           : 1553 (69.95%)
  Expectedly Failed:   52 (2.34%)
  Failed           :   72 (3.24%)
```

And here are the results when running E2E tests on PVC with `--offload-new-driver -foffload-lto=thin`, so enabling thinLTO:
```
Total Discovered Tests: 2220
  Unsupported      :  543 (24.46%)
  Passed           : 1466 (66.04%)
  Expectedly Failed:   52 (2.34%)
  Timed Out        :    1 (0.05%)
  Failed           :  158 (7.12%)
```

It seems we have a decent passrate, only 87 thinLTO specific E2E failures. I'm holding off investigating these until we finalize the architecture.
Once it's finalized, I'll make a design document to be submitted to this repository.

sarnex commented on libdevice/device.h:
```diff
@@ -17,9 +17,9 @@
 #ifdef __SYCL_DEVICE_ONLY__
-#define DEVICE_EXTERNAL SYCL_EXTERNAL __attribute__((weak))
+#define DEVICE_EXTERNAL SYCL_EXTERNAL
 #else // __SYCL_DEVICE_ONLY__
-#define DEVICE_EXTERNAL __attribute__((weak))
+#define DEVICE_EXTERNAL
 #endif // __SYCL_DEVICE_ONLY__
```
This is required to get libdevice working, see [here](https://github.com/intel/llvm/pull/14096). I'm looking for a better solution for this, I just kept this here in case anybody plans on trying the prototype.

bader replied:
I suppose importing devicelib symbols at compile step can be a solution (see https://github.com/intel/llvm/pull/15114).

On the other hand, I recall discussing the possibility of linking device libraries with upstream maintainers, who expressed a preference for shifting device library linking from the "compile" to the "link" step. It would be ideal if we could discover a solution that aligns with the long-term strategy of upstream and enables us to utilize the thinLTO framework for offload code linking, thereby avoiding the use of weak symbols.

asudarsa replied:
attention to @mdtoguchi who has been looking at importing devicelib at compile step from the SYCL perspective.
Point to note: During one of the LLVM community presentation, it was mentioned that they are trying to move importing devicelib to link time.

mdtoguchi replied:
As we already perform device library linking at link time we can consider abandoning the efforts to pull them into the compilation step.  My main concern with performing at the link step is the communication required from the driver to the `clang-linker-wrapper` informing which device libraries should be linked.  The less tie-in we have between the driver and the `clang-linker-wrapper` at link time, the better.  IMO, at the very least the linker wrapper should know a minimum default device libraries to link and any communication from the driver is manipulating that list.

bader replied:
I agree with @mdtoguchi. Unless user wants to change the names/location or disable linking of device libraries, driver should not have any logic to handle device code linking other than invoking `clang-linker-wrapper`. It makes sense to have driver options for additional configuration of device libraries, but driver's implementation should be just passing corresponding values to `clang-linker-wrapper` where these options should be processed.

MartinWehking replied:
That is very interesting.
While working on https://github.com/intel/llvm/pull/15114 I've been wondering whether there is a particular reason why we link against CUDA libdevice and libclc in the compile step, but also again in the link step.
Could I get some clarification on that?

bader replied:
> That is very interesting. While working on #15114 I've been wondering whether there is a particular reason why we link against CUDA libdevice and libclc in the compile step, but also again in the link step. Could I get some clarification on that?

@Naghasan, @npmiller, are you able to help here?

sarnex commented on llvm/lib/LTO/LTO.cpp:
```diff
@@ -1077,8 +1077,8 @@ Error LTO::addThinLTO(BitcodeModule BM, ArrayRef<InputFile::Symbol> Syms,
         ThinLTO.ModulesToCompile->insert({BM.getModuleIdentifier(), BM});
-        llvm::errs() << "[ThinLTO] Selecting " << BM.getModuleIdentifier()
-                     << " to compile\n";
+        LLVM_DEBUG(dbgs() << "[ThinLTO] Selecting " << BM.getModuleIdentifier()
```
By default if you set `ModulesToCompile`, the thinLTO infrastructure prints out the name of every match even in release builds. This seems like it should be debug only, so making that change here. This is not required for correctness.

bader replied:
We should propose this modification to the upstream.

> if you set `ModulesToCompile` in `LTOConfig` which I do

Why do we need to set `ModulesToCompile`?

sarnex replied:
We set it so that we don't try to codegen the devicelib modules, we only want to link in their functions used by users. Without setting this, every BC module added to the thinLTO infrastructure gets codegen run on it.

Good point, I'll make a note to make an upstream PR.

sarnex replied:
Merged upstream [here](https://github.com/llvm/llvm-project/commit/fedc7556ad5237cd11e29b8e70d412bdc39a4fa6)

bader commented:
> I'd like to get some architecture and code-organzation level comments on my SYCL thinLTO prototype.
> 
> The feature isn't ready for code-review, so I'd appreciate keeping the discussion somewhat high level.

Let's create a design doc for thinLTO feature. It will be easier to have this conversation when the design and open questions are kept in a separate document versioned by Git.

> The main idea of the change is as follows:
> 
> We hook into the existing LTO framework in `clang-linker-wrapper` by adding a callback that runs right before the backend runs (so right before the SPIR-V backend would be called).
> 
> Before the callback is run, function importing happens automatically inside the thinLTO framework where we link in the definitions for any functions referenced in this module but defined in another. When the callback is run, the module passed in should have everything linked in and be ready for processing.
> 
> In that callback, we do all the required processing of the SYCL module. In the current prototype, that includes: Generating the module properties Generating the module symbol table Calling the SPIR-V translator
> 
> In the future, we will also need to handle at least spec constants and internalization of non-entry points.

> 
> The final step of the callback, calling the SPIR-V translator, should be removed when the SPIR-V backend is ready.
> 
> One problem is that we need to store the module properties, symbol table, and SPIR-V translator output file path until a later time where we are ready to create the final images, which requires all of this information.
> 

I would prefer if we all "the required processing" in standard LLVM pipeline.

1. Generating the module properties.
2. Generating the module symbol table.
3. Calling the SPIR-V translator. SPIR-V translator will be replaced by SPIR-V backend.
4. Handle at least spec constants. I'm not 100% sure I fully understand the scope of work we need for "handling", but assuming that it's requires just transform LLVM IR, this should be done by injecting an additional pass to the standard pipeline. Hopefully this doesn't require storing additional information outside of LLVM IR module.
5. Internalization of non-entry points. AFAIK, "internalization" is very common task. Some sorts of internalizations are done by LTO framework itself, NVPTX and AMDGPU targets. A couple of links to the source code: https://github.com/intel/llvm/blob/ee975312fd443e19edc5e0674fe693cb660255ae/llvm/lib/Transforms/IPO/Internalize.cpp#L43-L52, https://github.com/intel/llvm/blob/ee975312fd443e19edc5e0674fe693cb660255ae/llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp#L213-L218.

We should consider doing steps 1 and 2 outside of LTO framework i.e. either before LTO or after. SPIR-V module has information about the properties and exported symbols. My understanding is that first two steps extracting this information from LLVM IR module into a data structure, which execution environment (in our case is SYCL runtime) uses to "handle" the optimized code at runtime (e.g. JIT compile and/or execute).

> We are also somewhat limited on what we can do to store the data, because the callbacks run on threads inside the LTO framework, and some data structures (like `StringSaver`) are not thread-safe. We need to do this information inside the LTO framework thread because that is the only point we have the IR with functions imported.
> 
> As a low-effort hack to get the prototype working, I'm just storing that information as `std::string`, which is pretty inefficient and gross.
> 
> I decided to store the information inside the `OffloadBinary`. Otherwise, we'd need to create and pass a SYCL-specific data structure all over the place, which seems bad. The current changes to `OffloadBinary` to add a temporary string vector is definitely disgusting, but I'm proposing we make some change to the class to allow us to store the information we need, but I don't have a good idea for a clean and thread-safe data structure to use. I don't have a better idea for a place to store all this information.
> 
> In terms of test results:
> 
> Here are the results when running E2E tests on PVC with `--offload-new-driver` only enabled:
> 
> ```
> Total Discovered Tests: 2220
>   Unsupported      :  543 (24.46%)
>   Passed           : 1553 (69.95%)
>   Expectedly Failed:   52 (2.34%)
>   Failed           :   72 (3.24%)
> ```
> 
> And here are the results when running E2E tests on PVC with `--offload-new-driver -foffload-lto=thin`, so enabling thinLTO:
> 
> ```
> Total Discovered Tests: 2220
>   Unsupported      :  543 (24.46%)
>   Passed           : 1466 (66.04%)
>   Expectedly Failed:   52 (2.34%)
>   Timed Out        :    1 (0.05%)
>   Failed           :  158 (7.12%)
> ```
> 
> It seems we have a decent passrate, only 87 thinLTO specific E2E failures. I'm holding off investigating these until we finalize the architecture. Once it's finalized, I'll make a design document to be submitted to this repository.

asudarsa commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -11231,8 +11231,12 @@ void LinkerWrapper::ConstructJob(Compilation &C, const JobAction &JA,
       CmdArgs.push_back(Args.MakeArgString("-sycl-thin-lto"));
+      // TODO: Pass the same value for this argument once we start using it
+      // for non-thinLTO.
+      CmdArgs.push_back(Args.MakeArgString("-sycl-module-split-mode=auto"));
```
I looked at the logic surounding 'sycl-module-split-mode' option. If this option is set (to any value), we use the SYCL split library. Otherwise, we call the SYCL post link tool. Given the current state of affairs, i would suggest that we do not set this option.
Thanks

asudarsa commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -1595,6 +1717,14 @@ std::unique_ptr<lto::LTO> createLTO(
   Conf.OptLevel = OptLevel[1] - '0';
   Conf.DefaultTriple = Triple.getTriple();
 
+  // We need to set up the backend to use thinLTO
+  // even if we don't actually use it, and there is no
+  // backend for the spir64 triple, so override it to
+  // the SPIR-V backlend.
+  // TODO: Remove once SYCL uses the SPIR-V backend.
```
+1 on this. I also briefly tried to use backend for 'linking' LLVM IR files. I am planning to revisit that PR (https://github.com/intel/llvm/pull/13395) soon. 
Also, eventually, we will use SPIRV64 (instead of SPIR64) as the target triple for SYCL purposes.
This change is agreeable as a stop gap. I think.
Adding @Naghasan for comments.

Thanks

sarnex replied:
Sounds good to me, the fact we can't get anywhere with the `spir` triples is a bit frustrating.

asudarsa commented:
> > I'd like to get some architecture and code-organzation level comments on my SYCL thinLTO prototype.
> > The feature isn't ready for code-review, so I'd appreciate keeping the discussion somewhat high level.
> 
> Let's create a design doc for thinLTO feature. It will be easier to have this conversation when the design and open questions are kept in a separate document versioned by Git.
> 
> > The main idea of the change is as follows:
> > We hook into the existing LTO framework in `clang-linker-wrapper` by adding a callback that runs right before the backend runs (so right before the SPIR-V backend would be called).
> > Before the callback is run, function importing happens automatically inside the thinLTO framework where we link in the definitions for any functions referenced in this module but defined in another. When the callback is run, the module passed in should have everything linked in and be ready for processing.
> > In that callback, we do all the required processing of the SYCL module. In the current prototype, that includes: Generating the module properties Generating the module symbol table Calling the SPIR-V translator
> > In the future, we will also need to handle at least spec constants and internalization of non-entry points.
> 
> > The final step of the callback, calling the SPIR-V translator, should be removed when the SPIR-V backend is ready.
> > One problem is that we need to store the module properties, symbol table, and SPIR-V translator output file path until a later time where we are ready to create the final images, which requires all of this information.
> 
> I would prefer if we all "the required processing" in standard LLVM pipeline.
> 
> 1. Generating the module properties.
> 2. Generating the module symbol table.
> 3. Calling the SPIR-V translator. SPIR-V translator will be replaced by SPIR-V backend.
> 4. Handle at least spec constants. I'm not 100% sure I fully understand the scope of work we need for "handling", but assuming that it's requires just transform LLVM IR, this should be done by injecting an additional pass to the standard pipeline. Hopefully this doesn't require storing additional information outside of LLVM IR module.
> 5. Internalization of non-entry points. AFAIK, "internalization" is very common task. Some sorts of internalizations are done by LTO framework itself, NVPTX and AMDGPU targets. A couple of links to the source code: https://github.com/intel/llvm/blob/ee975312fd443e19edc5e0674fe693cb660255ae/llvm/lib/Transforms/IPO/Internalize.cpp#L43-L52
>    , https://github.com/intel/llvm/blob/ee975312fd443e19edc5e0674fe693cb660255ae/llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp#L213-L218
>    .
> 
> We should consider doing steps 1 and 2 outside of LTO framework i.e. either before LTO or after. SPIR-V module has information about the properties and exported symbols. My understanding is that first two steps extracting this information from LLVM IR module into a data structure, which execution environment (in our case is SYCL runtime) uses to "handle" the optimized code at runtime (e.g. JIT compile and/or execute).
> 
> > We are also somewhat limited on what we can do to store the data, because the callbacks run on threads inside the LTO framework, and some data structures (like `StringSaver`) are not thread-safe. We need to do this information inside the LTO framework thread because that is the only point we have the IR with functions imported.
> > As a low-effort hack to get the prototype working, I'm just storing that information as `std::string`, which is pretty inefficient and gross.
> > I decided to store the information inside the `OffloadBinary`. Otherwise, we'd need to create and pass a SYCL-specific data structure all over the place, which seems bad. The current changes to `OffloadBinary` to add a temporary string vector is definitely disgusting, but I'm proposing we make some change to the class to allow us to store the information we need, but I don't have a good idea for a clean and thread-safe data structure to use. I don't have a better idea for a place to store all this information.
> > In terms of test results:
> > Here are the results when running E2E tests on PVC with `--offload-new-driver` only enabled:
> > ```
> > Total Discovered Tests: 2220
> >   Unsupported      :  543 (24.46%)
> >   Passed           : 1553 (69.95%)
> >   Expectedly Failed:   52 (2.34%)
> >   Failed           :   72 (3.24%)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And here are the results when running E2E tests on PVC with `--offload-new-driver -foffload-lto=thin`, so enabling thinLTO:
> > ```
> > Total Discovered Tests: 2220
> >   Unsupported      :  543 (24.46%)
> >   Passed           : 1466 (66.04%)
> >   Expectedly Failed:   52 (2.34%)
> >   Timed Out        :    1 (0.05%)
> >   Failed           :  158 (7.12%)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It seems we have a decent passrate, only 87 thinLTO specific E2E failures. I'm holding off investigating these until we finalize the architecture. Once it's finalized, I'll make a design document to be submitted to this repository.

@bader, can you please clarify..'standard LLVM pipeline'? Sorry if the question is too naive. Wanted to check before adding my comments here. Thanks

asudarsa commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -590,6 +591,58 @@ static Error getSYCLDeviceLibs(SmallVector<std::string, 16> &DeviceLibFiles,
   return Error::success();
 }
 
+static Error getDeviceLibsForLTO(SmallVector<OffloadFile> &DeviceLibs,
```
Can we just call this getDeviceLibs or some generic name and use it in https://github.com/intel/llvm/blob/70ee0e04c960c06fc69323f7597232700cb34957/clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp#L1364 as well?
Is that possible?
Thanks

asudarsa replied:
Marking this as resolved...as this is a RFC PR and I do not think we need to address code improvements here.

asudarsa commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -941,6 +1016,13 @@ static Expected<StringRef> runLLVMToSPIRVTranslation(StringRef File,
   return *TempFileOrErr;
 }
 
+static Expected<StringRef> runLLVMToSPIRVTranslation(StringRef File,
```
Just a curiousity question. Why was this modification required?
Seems like functionally equivalent to the old flow and the way it's invoked for thinLTO also seems similar.

Thanks

sarnex replied:
It was because I was hitting multithreading issues, we can't use `StringRefs` or `Args` as we want on any thread, so I computed the args before the thinLTO threads are created to work around the issue. Definitely a hack and will require a real solution at some point.

sarnex replied:
That definitely sounds like a good idea to me. @asudarsa, you mentioned we were seeing 95% pass rate on E2E tests with optimization enabled with the SPIR-V backend. Do you agree with dropping support for llvm-spirv in thinLTO and only supporting the backend?

asudarsa replied:
+1 on that. I am planning to call for a quick meeting before we do this. I am hoping to use SPIR-V backend for new offload model while we continue to use SPIR-V translator in the old offloading model. This will make the switch to NOT impact our customers right away.

asudarsa commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -1750,11 +1982,11 @@ Error linkBitcodeFiles(SmallVectorImpl<OffloadFile> &InputFiles,
-                            Args.hasArg(OPT_clang_backend)
-                        ? createLTO(Args, Features, OutputBitcode)
-                        : createLTO(Args, Features);
+  auto LTOBackend =
```
This does not look like a SYCL specific change (atleast on face value). Should this go upstream?

sarnex replied:
It is not technically SYCL specific, but in this PR SYCL is the only client using the new arguments.

bader commented:
> @bader, can you please clarify..'standard LLVM pipeline'? Sorry if the question is too naive. Wanted to check before adding my comments here. Thanks

I mean the pipeline built here: 
https://github.com/intel/llvm/blob/283073a4f21aa44203fd49e25b4ff0379c91b29a/llvm/include/llvm/Passes/PassBuilder.h#L264-L282

sarnex commented:
@bader 

Thanks for the detailed feedback.

> Let's create a design doc for thinLTO feature. It will be easier to have this conversation when the design and open questions are kept in a separate document versioned by Git.

Okay, will work on it.

> We should consider doing steps 1 and 2 outside of LTO framework i.e. either before LTO or after. SPIR-V module has information about the properties and exported symbols. My understanding is that first two steps extracting this information from LLVM IR module into a data structure, which execution environment (in our case is SYCL runtime) uses to "handle" the optimized code at runtime (e.g. JIT compile and/or execute).

Correct, we need to extract this information into a data structure which will be finally used by the runtime for many things. I originally wanted to do it inside the LTO framework so that we wouldn't have to look at the IR at all outside of the framework for performance reasons (thinLTO uses `BitcodeModule` which is lazy-load), but since we don't have one single module anymore and we shouldn't need functions linked in to compute this infomration, I think it's reasonable to do it outside the framework. Let me try this out.

> Handle at least spec constants. I'm not 100% sure I fully understand the scope of work we need for "handling", but assuming that it's requires just transform LLVM IR, this should be done by injecting an additional pass to the standard pipeline. Hopefully this doesn't require storing additional information outside of LLVM IR module.

The problem with spec constants is that based on comments from@AlexeySachkov, we need the entire IR for the module to accurately compute them, stubs for functions in other TUs won't cut it. The only time that is true with thinLTO is inside the thinLTO framework after function importing. @AlexeySachkov Let me know if I'm wrong about this requirement.

> Internalization of non-entry points. AFAIK, "internalization" is very common task. Some sorts of internalizations are done by LTO framework itself, NVPTX and AMDGPU targets. A couple of links to the source code: 

Yeah really good point here, we should just be able to make this a pass that is called by LTO. Let me work on this.

If these ideas work I think we can clean this up a lot, thanks again. Will report back when I have more info.

AlexeySachkov commented:
> One problem is that we need to store the module properties, symbol table, and SPIR-V translator output file path until a later time where we are ready to create the final images, which requires all of this information.

Why do we need SPIR-V translator output path file to be generated at this stage? Why can't it be generated directly when we call the translator? Why do we even need a file here? The translator has library interface and if it lacks API that works on memory buffers, then let's just add such API in there so we don't have to deal with files at all.

> We are also somewhat limited on what we can do to store the data, because the callbacks run on threads inside the LTO framework, and some data structures (like StringSaver) are not thread-safe. We need to do this information inside the LTO framework thread because that is the only point we have the IR with functions imported.

Could you please clarify why do we have a problem here at all? My assumption is that every thinLTO thread works on a separate translation unit and therefore produces a separate module and `OffloadBinary` object, i.e. it doesn't touch properties of other modules.

In any case, one more thing that we can do is to slightly refactor how we deal with those properties to delay recording them into `OffloadBinary`. What we can do is to record those properties into module named metadata at first and then extract when we need to write them down into the final device image data structure. This is essentially how `sycl-post-link` operates and my guess is that there should be a point in time where we perform a final wrapping of a single module - at that point we can read metadata and copy/move it to `OffloadBinary`.

> The problem with spec constants is that based on comments from @AlexeySachkov, we need the entire IR for the module to accurately do the processing, stubs for functions in other TUs won't cut it. The only time that is true with thinLTO is inside the thinLTO framework after function importing. @AlexeySachkov Let me know if I'm wrong about this requirement.

That is correct, spec constants won't work if transformation is done before link phase. Reason for that is the feature design: in SYCL spec constants are identified by non-type template parameters (somewhat similar to kernels) and under the hood we use strings emitted for us by a special built-in similar to unique stable name. The pass then converts those strings into integers and records a mapping string -> integer. That is needed because at SPIR-V level spec constants are identified by their IDs. If we have two translation units which both use spec constants and they are linked together, then we can't perform string to ID replacement before link stage. Simply because it can happen that we will record two different spec constants to have the same integer ID. Theoretically, we could do a renumeration at this point, but there is also a case with default values where global (module-wide) order of spec constants handling matters and we won't be able to easily (or at all) re-enumerate them in that case, it is different from SPIR-V (native) path.

Note that spec constants is only one example, but there are (and will be) likely more passes that require fully linked module to work correctly.

> Generating the module properties.

@bader, while we can _prepare_ properties in standard LLVM pipeline, we can't generate their final representation within the pipeline, simply because they have to be stored within a wrapped object that is generated by `clang-linker-wrapper` and not by the optimization pipeline.

> > Internalization of non-entry points. AFAIK, "internalization" is very common task. Some sorts of internalizations are done by LTO framework itself, NVPTX and AMDGPU targets. A couple of links to the source code:
> Yeah really good point here, we should just be able to make/modify a pass that is called by LTO. Let me work on this.

Note that we already use standard LLVM pass for internalization and we launch it always, just with different "filters", see #14976

sarnex commented:
@AlexeySachkov Thanks for the feedback.

> Why do we need SPIR-V translator output path file to be generated at this stage? Why can't it be generated directly when we call the translator? Why do we even need a file here? The translator has library interface and if it lacks API that works on memory buffers, then let's just add such API in there so we don't have to deal with files at all.

We don't need a file, but we need to somehow get the output of the SPIR-V translation process outside of the thinLTO thread and the mapping from input file vector to output file vector is not 1-to-1. I have an idea of how we can make that somewhat cleaner as well as address this feedback to use the library and not use files. Will work on this and propose it here.

> Could you please clarify why do we have a problem here at all? My assumption is that every thinLTO thread works on a separate translation unit and therefore produces a separate module and OffloadBinary object, i.e. it doesn't touch properties of other modules.

> In any case, one more thing that we can do is to slightly refactor how we deal with those properties to delay recording them into OffloadBinary. What we can do is to record those properties into module named metadata at first and then extract when we need to write them down into the final device image data structure. This is essentially how sycl-post-link operates and my guess is that there should be a point in time where we perform a final wrapping of a single module - at that point we can read metadata and copy/move it to OffloadBinary.

The main problem I noticed was trying to save StringRefs created on the thinLTO thread. To save StringRefs in general, we need StringSaver, but StringSaver is not threadsafe. I'm somewhat embaressed to say I didn't investigate the issue to the root-cause but just worked around it with the std::string hack. I think I can remove the hack while addressing feedback given here.

Saving it as metadata is a good idea, will try that. Not ideal, but should be a lot cleaner on the `clang-linker-wrapper` code.

> That is correct, spec constants won't work if transformation is done before link phase. Reason for that is the feature design: in SYCL spec constants are identified by non-type template parameters (somewhat similar to kernels) and under the hood we use strings emitted for us by a special built-in similar to unique stable name. The pass then converts those strings into integers and records a mapping string -> integer. That is needed because at SPIR-V level spec constants are identified by their IDs. If we have two translation units which both use spec constants and they are linked together, then we can't perform string to ID replacement before link stage. Simply because it can happen that we will record two different spec constants to have the same integer ID. Theoretically, we could do a renumeration at this point, but there is also a case with default values where global (module-wide) order of spec constants handling matters and we won't be able to easily (or at all) re-enumerate them in that case, it is different from SPIR-V (native) path.

> Note that spec constants is only one example, but there are (and will be) likely more passes that require fully linked module to work correctly.

My plan is to add passes to the thinLTO pipeline of passes to run using the `PreCodeGenPassesHook` field of `LTOConfig`, but unfortunately we need to use the old passes manager because that's all it supported. It seems it uses the old pass manager because a lot of the codegen passes still do. We should be able to add passes as needed here, and if the legacy pass manager stuff becomes too annoying we can make a single pass for us to call that calls everything we need.

Overall, thanks a lot for the feedback, you've given me some ideas I should be able to use to clean up the implementation a lot.

AlexeySachkov commented:
> My plan is to add passes to the thinLTO pipeline of passes to run using the PreCodeGenPassesHook field of LTOConfig, but unfortunately we need to use the old passes manager because that's all it supports. It seems it uses the old pass manager because a lot of the codegen passes still do. We should be able to add passes as needed here, and if the legacy pass manager stuff becomes too annoying we can make a single legacy pass for us to call that calls everything we need.

I suppose that it should be possible to wrap "new" passed into "legacy" interface so we can run them in both pipelines without having to duplicate them completely. At least the vice-verse wrapping was possible and that's what we used during transition to the new pass manager.

bader commented:
> Note that spec constants is only one example, but there are (and will be) likely more passes that require fully linked module to work correctly.

In my opinion, this is a significant disadvantage of the design and should be avoided. There is an extension proposal with alternative API providing similar functionality w/o design limitations of spec constants API.

sarnex commented:
> To clarify: by fully linked I only mean that further linking of two modules that have undergone spec constants processing can't be performed, because spec constants will be broken. It doesn't really require to link all kernels together. We just can't link them together afterwards.

Do you mean it would be semantically correct to run spec constants processing individually on each module, as long as we only do it once per output file? So if we had 2 obj files each with 2 splits, we'll have 4 output modules, so could we run the pass at any point in time on each module, as long as we only do it once? We wouldn't even need to run it inside thinLTO then.

AlexeySachkov commented:
> Do you mean it would be semantically correct to run spec constants processing individually on each module, as long as we only do it once per output file? So if we had 2 cpp files input to clang each with 2 splits assuming arbitrary cross-TU dependencies, we'll have 4 output modules, so could we run the pass at any point in time on each module, as long as we only do it once?

Right.

> We wouldn't even need to run it inside thinLTO then, we could do it before or after.

I'm not sure about before, because `thinLTO` may bring a function that uses spec constants to a kernel that also uses spec constants which will lead to integer IDs clash. After is perfectly fine, though. That's how it is currently implemented except that we have `full` instead of thing and no optimizations :)

sarnex commented:
> I'm not sure about before, because thinLTO may bring a function that uses spec constants to a kernel that also uses spec constants which will lead to integer IDs clash.

> UPD: if we assume that thinLTO doesn't change a module, i.e. it won't bring anything new to it, then yes, we can run a pass even before thinLTO

Sorry, I edited my question a ton. It depends on the defintion of change, thinLTO will definitely change the module by importing function definitions. I think just using the `PreCodeGenPassesHook` lambda to call it from thinLTO where everything is linked in is the cleanest anyway.

AlexeySachkov commented:
> Sorry, I edited my question a ton. It depends on the defintion of change, thinLTO will definitely change the module by importing function definitions and running optimizations.

Optimizations are not a concern, importing of new functions is. We can't link two modules which were processed by spec constants pass, so it has to be run _after_ importing, but regardless of optimizations.

sarnex commented:
@bader @asudarsa @AlexeySachkov @RaviNarayanaswamy @mdtoguchi @maksimsab 

Hey all, I just pushed a new commit with a design doc and updated prototype implementation based on the feedback. Please take a look when you have a chance. Thanks!

AlexeySachkov commented on libdevice/fallback-cassert.cpp:
```diff
@@ -114,9 +114,4 @@ DEVICE_EXTERN_C void __devicelib_assert_fail(const char *expr, const char *file,
   __assertfail(expr, file, line, func, 1);
 }
 
-DEVICE_EXTERN_C void _wassert(const char *_Message, const char *_File,
```
I feel like this is a change that can be merged and submitted separately. `_wassert` is a wrapper for MSVC's `assert` implementation to redirect it to ours, so it really shouldn't be implemented in `fallback` library

sarnex replied:
I don't even know if it's correct, I just hit a build error on windows about `_wassert` defined twice, probably it works today because they're all weak symbols but I remove that as part of this PR.

AlexeySachkov commented on llvm/lib/SYCLLowerIR/SYCLLinkedModuleProcessor.cpp:
```diff
@@ -0,0 +1,45 @@
+  }
+
+  bool runOnModule(Module &M) override {
+    // TODO: determine if we need to run other passes
```
If I understand correctly, that's an equivalent of what's being run by `sycl-post-link` after device code split is performed. If so, then we have the following other transformations applied at this stage:
- ESIMD handling, which includes some special module fixup for `invoke_simd`, as well as potential additional split by ESIMD followed up by optional linking that back
- Generation of a separate device image with default values of spec constants

If we also taking about what happens after `llvm-link` but before device code split, then it is also:
- Something about `invoke_simd`
- Sanitizer-related passes
- Joint matrix passes

sarnex replied:
So when we do early splitting in `-c` we actually run sycl-post-link in full, including all those passes. So in that case, we only need to run passes here that need the fully linked module. If we decide to change the design such that we do only split in `-c` but no passes, then we would need every pass that sycl-post-link runs. In the current implementation ~2100/2000 E2E tests are passing, so it seems most passes don't need the full module, at least for the test cases we have.

AlexeySachkov replied:
> In the current implementation ~2100/2000 E2E tests are passing, so it seems most passes don't need the full module, at least for the test cases we have.

I believe that most of E2E are single-file tests with no `SYCL_EXTERNAL` dependencies. Even SYCL-CTS won't help you here. I suppose that we need more or less real-life applications here to be sure and gather more data if we need it

bader replied:
I suggest adding tests similar to `sycl/test-e2e/Basic/multisource.cpp`.

AlexeySachkov commented on sycl/doc/design/images/ThinLTOCommunityFlow.svg:
I don't know what is our general stand on this, but GH natively supports rendering of diagrams from markdown using some `mermaid` library. See the docs here: https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams#creating-mermaid-diagrams
Full docs here: https://mermaid.js.org/intro/ (and you can also use live editor there to test stuff quicker than commit-push-refresh web page)

Tagging @intel/dpcpp-doc-reviewers here for opinion if we are fine to use that functionality. To me, that looks interesting: it should be more or less readable in raw form and it is way easier to edit and update than `.svg` files added to the repo. So I'm **for** allowing to us that functionality in our docs (and actually using it instead of accepting `.svg` files)

sarnex replied:
If `mermaid` allows creating all of the diagrams we need to, I'm definitely in favor of it, using `.svg` is a nightmare, at least for me.

bader replied:
No objections from my side. I see a few uses of `mermaid` diagrams in our documentation already:
- https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_forward_progress.asciidoc#execution-model
- https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc#751-graph-state
- https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc#761-queue-state

sarnex replied:
Is it okay if I did this in a separate PR? It took me forever to make the current images and don't want to spend possibly forever making new ones :)

AlexeySachkov reviewed (commented):
A very brief skim over, I haven't read the new doc yet

bader commented:
@sarnex, I've made minor editorial changes, but unfortunately, I can't push them to your branch. Your repository doesn't allow me to create a pull request to your private branch. If you are okay with the changes from [these two commits](https://github.com/bader/llvm/compare/43df8c948764749a7a40421ddfc14acb69721f6e...bader:llvm:thinlto), please, cherry-pick them to your private branch.

bader commented on sycl/doc/design/ThinLTO.md:
```diff
@@ -0,0 +1,147 @@
+part of device linkage. At least for AMD devices, this is not a requirement as 
+`lld` is used for the final link which can resolve any unresolved symbols. 
+`-fno-gpu-rdc` is default for AMD, so in that case it can call `lld` during 
+compile, but if `-fno-gpu-rdc` is passed, the lld call happens as part of 
```
```suggestion
compile, but if `-fgpu-rdc` is passed, the lld call happens as part of 
```

sarnex replied:
pulled your branch, thanks!

bader reviewed (commented):
Documentation part review.

sarnex commented on sycl/doc/design/ThinLTO.md:
```diff
@@ -0,0 +1,147 @@
+libraries) from other object files.
+
+We rely on the assumption that all function defintions matching a declaration 
+will be the same and we can let ThinLTO pull in any one.
```
I think it depends what the original IR linkage is. if the original IR is `linkonce_odr` or something similar I think yes, but I don't know if we can guarantee every SYCL function will have that linkage (at least for libdevice it not this way in syclos HEAD)

sarnex commented on sycl/doc/design/ThinLTO.md:
```diff
@@ -0,0 +1,147 @@
+ThinLTO infrastruction during the processing of `tu_bar` is valid.
+
+As a result of running early device code split, the fat object file generated 
+as part of device compilation may contain multiple device code images.
```
After split, the linkage should be the same as it was before split.

After ThinLTO runs, it could have the same linkage as after splitting or it could be internalized (not yet implemented)

I don't think there is any way to get multiple kernel definitions in a way that isn't already possible with splitting. Maybe @AlexeySachkov has an idea.

bader replied:
> After split, the linkage should be the same as it was before split.

I'm not 100% sure, but this might cause "multiple definition" problem. `tu_foo` has only one `foo` definition so using `external` is fine (assuming that all other modules referencing `foo` use different linkage types), but after split we will have `foo` defined in multiple modules. I'm not sure if LLVM allows linking modules where `foo` will have `external` linkage type.

AlexeySachkov replied:
> I don't think there is any way to get multiple kernel definitions in a way that isn't already possible with splitting. Maybe @AlexeySachkov has an idea.

We shouldn't ever duplicate kernels in our compiler stack, that won't be properly handled at RT for multiple reasons.

bader commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -2365,9 +2614,21 @@ Expected<SmallVector<StringRef>> linkAndWrapDeviceFiles(
+    llvm::Triple Triple(LinkerArgs.getLastArgValue(OPT_triple_EQ));
+    if (!linkerSupportsLTO(LinkerArgs)) {
+      if (sycl::isSYCLThinLTO(Args, Triple)) {
+        // For thinLTO, we consider device libs as normal compiler input
+        // and add them to the files to be processed by the LTO backend.
```
This is something we should review with the community. As of now, they import some parts of GPU devicelib at compile time (i.e. LLVM IR). Considering community interest in moving this action to the link step, we better align with them.

Looking at some patches in upstream, I feel like some calls to GPU libraries are handled at link step (e.g. https://github.com/llvm/llvm-project/pull/98512).

sarnex commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -1814,18 +1951,16 @@ std::unique_ptr<lto::LTO> createLTO(
-      // properties and the module symbol table, convert to SPV (using the
-      // translator for now) and save required information for binary created
-      // inside the OffloadFile.
+    auto SpecConstArg = sycl::areSpecConstsSupported(Args, Triple)
+                            ? SpecConstantsPass::HandlingMode::native
+                            : SpecConstantsPass::HandlingMode::emulation;
```
@AlexeySachkov Any reservations about this change to `SpecConstantsPass`? Happy to make the change myself.

AlexeySachkov replied:
We also have a 3rd mode which is to replace spec constants with default values. That 3rd mode we use alongside `emulation`, i.e. we may invoke the pass twice (not in a row, but "side by side") to produce two different modules with different handling of spec constants.

So, whilst it should be possible to select between `native` vs `emulation` within the pass based on the module alone, I'm not that sure about `default` mode.

sarnex commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -1835,43 +1970,18 @@ std::unique_ptr<lto::LTO> createLTO(
-      OffloadF.getBinary()->addTmpString(ModuleSyms);
-      // TODO: Use SPIR-V backend instead of SPIR-V translator once the backend
-      // is mature.
+      // Use the legacy PM because eventually we will use the
+      // PreCodeGenPassesHook field of LTOConfig which requires the legacy PM.
+      legacy::PassManager PM;
```
Since we return `false` from `PreCodeGenPassesHook` thus telling the LTO framework to not continue, `PostOptModuleHook` is not invoked. Once we use the SPIR-V backed, this should definitively be inside `PostOptModuleHook.

bader replied:
> Since we return `false` from `PreCodeGenPassesHook` thus telling the LTO framework to not continue, `PostOptModuleHook` is not invoked.

According to my understanding LTOBackend.cpp code, the `PostOptModuleHook` is invoked before the `PreCodeGenPassesHook`. 

I think we need to separate codegen to SPIR-V and running "post-link" optimization passes. SYCLLinkedModuleProcessorPass can be added even into `PostImportModuleHook` as we want to run standard llvm optimizations **after** SYCL passes handle spec constants, joint matrix, etc.

sarnex replied:
Sorry, I typo'd it, the correct response is

Since we return false from `PreCodeGenModuleHook`  thus telling the LTO framework to not continue, PreCodeGenPassesHook is not invoked.

But sure, I'll move to the SPV backend and apply this feedback

sarnex commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -1450,6 +1458,135 @@ static Expected<StringRef> linkDevice(ArrayRef<StringRef> InputFiles,
   return *DeviceLinkedFile;
 }
 
+static bool isStaticArchiveFile(const StringRef Filename) {
```
None of this logic is part of thinLTO, it's old code I added back to allow using old objects in `clang-linker-wrapper`, it was originally written by Arvind I believe earlier this year. Sorry if it showed up in the changes here, probably it was from a merge commit. We plan to remove that code in the near future, so I'm not sure we need to change the implementation as we already have tests for it.

bader replied:
> We plan to remove that code in the near future, so I'm not sure we need to change the implementation as we already have tests for it.

👍 
Please, make sure we have a tracker for this work.

sarnex replied:
Made an internal tracker and put it in the new offloading model epic.

sarnex commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -1450,6 +1458,135 @@ static Expected<StringRef> linkDevice(ArrayRef<StringRef> InputFiles,
+
+// This routine is used to run the clang-offload-bundler tool and unbundle
+// device inputs that have been created with an older compiler where the
+// device object is bundled into a host object.
```
Correct, this code will be removed, it's just to help us switch the default in `intel/llvm`.

FYI @asudarsa

sarnex commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -635,6 +689,32 @@ static Expected<StringRef> convertSPIRVToIR(StringRef Filename,
+}
+
+bool isSYCLThinLTO(const ArgList &Args, const llvm::Triple Triple) {
+  // TODO: Support CUDA/HIP
```
Yes, sorry I meant using SYCL with NVPTX and AMDGPU targets. I'll update the comment.

sarnex commented:
@bader For some reason I can't reply inline

``
Considering that DPC++ passes most of end-to-end tests with SPIR-V backend, should we drop the translator support in thinLTO flow? I suppose we won't hit multithreading issues with SPIR-V backend.
```

That definitely sounds like a good idea to me. @asudarsa, you mentioned we were seeing 95% pass rate on E2E tests with optimization enabled with the SPIR-V backend. Do you agree with dropping support for `llvm-spirv` in thinLTO and only supporting the backend?

sarnex commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -671,10 +751,7 @@ getTripleBasedSYCLPostLinkOpts(const ArgList &Args,
   // because it only increases amount of code for device compiler to handle,
   // without any actual benefits.
   // TODO: Try to extend this feature for non-Intel GPUs.
```
good catch, will fix. thanks!

sarnex commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -1774,11 +1912,21 @@ std::unique_ptr<lto::LTO> createLTO(
+  auto PreCodeGenSaveTemps = [=](size_t Task, const Module &M) {
+    std::string File =
+        !Task ? TempName + ".postopt.bc"
+              : TempName + "." + std::to_string(Task) + ".postopt.bc";
```
will do, thx!

sarnex commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -600,6 +602,58 @@ static Error getSYCLDeviceLibs(SmallVector<std::string, 16> &DeviceLibFiles,
+    if (Error Err =
+            extractOffloadBinaries(Buffer->getMemBufferRef(), Candidates))
+      return Err;
+    for (OffloadFile &OffF : Candidates)
+      if (llvm::Triple(OffF.getBinary()->getTriple()) == Triple)
+        DeviceLibs.emplace_back(std::move(OffF));
```
we need to `std::move` here, will `copy_if` do that?

bader replied:
I'm not sure, so let's keep your version.

sarnex commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -600,6 +602,58 @@ static Error getSYCLDeviceLibs(SmallVector<std::string, 16> &DeviceLibFiles,
+            "Number of device library files cannot be zero.");
+      for (StringRef Val : A->getValues()) {
+        SmallString<128> LibName(Val);
+        if (llvm::sys::fs::exists(LibName)) {
+          if (auto Err = processFile(LibName))
+            return Err;
+        } else
+          return createStringError(
+              inconvertibleErrorCode(),
+              std::string(LibName) +
+                  " SYCL device library file for NVPTX is not found.");
```
yeah this is better, thanks!

sarnex commented on clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp:
```diff
@@ -635,6 +689,32 @@ static Expected<StringRef> convertSPIRVToIR(StringRef Filename,
+  // TODO: Try to extend this feature for non-Intel GPUs.
+  return (!Args.hasFlag(OPT_no_sycl_remove_unused_external_funcs,
+                        OPT_sycl_remove_unused_external_funcs, false) &&
+          !SYCLNativeCPU) &&
+         !Triple.isNVPTX() && !Triple.isAMDGPU();
```
good catch, i think yes, let me try.

bader commented:
> For some reason I can't reply inline

You can reply inline here: https://github.com/intel/llvm/pull/15083#discussion_r1769334803. When you publish a review with new discussions (1) and replies to already existing discussions (2), the replies (2) are posted in the original discussion where you can answer inline and duplicated in the review w/o the inline comment option.

sarnex commented:
Fixed, thx. Github UI confused me.

sarnex commented:
I'm planning to wait until we implement the new SYCL offloading design before picking this back up, as none of the code I added here will be run at all in that flow.

AlexeySachkov commented on sycl/doc/design/ThinLTO.md:
```diff
@@ -0,0 +1,149 @@
+## Device code compilation time changes
+
+Most of the changes for ThinLTO occur during device link time, however there is
+one major change during compilation (-c) time: we now run device code split
```
```suggestion
one major change during compilation (`-c`) time: we now run device code split
```

AlexeySachkov commented on sycl/doc/design/ThinLTO.md:
```diff
@@ -0,0 +1,149 @@
+but linking happens once total for the application. Device code split is
+currently a common source of performance issues.
+
+Splitting early means that the resulting IR after splitting is not complete, it
```
I don't think that early split leads to incomplete modules. Lack of linking leads to incomplete modules, i.e. a single translation unit with a call to undefined `SYCL_EXTERNAL` function is incomplete regardless of whether we split it or not.

If it was complete, then current version of the splitting mechanism won't break it, i.e. every resulting module would still be complete (at the cost of code duplication). However, I'm not 100% about the latter, if shared libraries are involved, that behavior definitely changes.

sarnex replied:
sorry, you're right, let me reword it

AlexeySachkov commented on sycl/doc/design/ThinLTO.md:
```diff
@@ -0,0 +1,149 @@
+
+For example, let's start with user device code that defines a `SYCL_EXTERNAL`
+function `foo` in translation unit `tu_foo`. There is also another translation
+unit `tu_bar` that references `foo`. During the early device code splitting run
+of `tu_foo`, we may find that more than one of the resultant device images
+contain a definition for `foo`.
```
Should we clarify the reason why `foo` could be duplicated? I'm not sure if that's obvious for a person who is not familiar with device code split details.

Also, we have a flag (originally introduced to support shared libraries) which allows to disable cloning of functions which are shared between device images produced by code splitting. We could probably use that mode for thinLTO to avoid unnecessary duplicates.

sarnex replied:
will add this, thanks

AlexeySachkov commented on sycl/doc/design/ThinLTO.md:
```diff
@@ -0,0 +1,149 @@
+
+## Device code link time changes
+
+Before we go into the link time changes for SYCL, let's understand the device
+linking flow for AMD/NVIDIA devices:
```
I read this phrase as **SYCL** linking flow for AMD/NVIDIA devices, but if that reading is correct, then requirement (2) looks weird, because it is not dependent on a device.

Perhaps this phrase was referring to upstream handling of OpenMP offloading?

sarnex replied:
yeah it's for omp, will reword. thanks

AlexeySachkov commented on sycl/doc/design/ThinLTO.md:
```diff
@@ -0,0 +1,149 @@
+
+The effect of requirement 1) is that instead of letting ThinLTO call the SPIR-V
+backend, we add a callback that runs right before CodeGen would run. In that
+callback, we call the SPIR-V translator and store the resultant file path for
+use later, and we instruct the ThinLTO framework to not perform CodeGen.
```
Note: we should be able to call the translator as a library and store the final binary (in form of a `MemoryBuffer`, I assume) instead of file path. The former is supposedly more in line with what happens in the upstream.

AlexeySachkov commented on sycl/doc/design/ThinLTO.md:
```diff
@@ -0,0 +1,149 @@
+
+1) We are relying on the external `llvm-spirv` tool instead of the SPIR-V
+backend. We could slightly improve this issue by using a library call to the
+SPIR-V translator instead of the tool, however the library API requires setting
+up an object to represent the arguments while we only have strings, and it's
+non-trivial to parse the strings to figure out how to create the argument
```
Am I right that you are talking about extensions here?

I don't think that it is necessarily non-trivial. If what we get from clang driver is what we should pass to `llvm-spirv`, then we can simply copy an existing function from `llvm-spirv` which does this parsing:

https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/main/tools/llvm-spirv/llvm-spirv.cpp#L516

sarnex replied:
it's not just that, if it were the extensions it would be not that bad.

let's consider the flag `llvm-spirv` flag `spirv-max-version`. This is a `cl::opt` global variable and is set in `llvm-spirv`'s main by parsing argv as is standard.

In `clang-linker-wrapper`, this is just 
If we want to set the same flag, we would need to somehow get that same global variable set, either by externing it and seting it ourselves (which means we need to parse the strings from the driver and map them to a global variable as well as all the enum mappings for the options). Doing that for all options is what I meant by non-trivial. 

If I missed some obvious easy way of doing it let me know

AlexeySachkov replied:
> let's consider the flag `llvm-spirv` flag `--spirv-max-version`. This is a `cl::opt` global variable `MaxSPIRVVersion` and is set in `llvm-spirv`'s main by parsing argv as is standard.

Do we have a public `clang` driver flag which allows to set maximum SPIR-V version? If not, do we pass different values of that option to the translator based on some other flags?

I guess my high-level point is: do we really have to pass all those flags through the linker-wrapper, or can we just come up with their values _within_ clang-linker wrapper based on the module itself?
That point of view stems from the fact that I don't really know which flags to the translator are currently "dynamic" and my assumption (which could be wrong one) that we don't have that many "dynamic" options which would actually require any handling.

sarnex replied:
Sorry, ` --spirv-max-version` was just a random option I picked, that one in particular might always have the same value. But even if it does, we need to implement some way to get that global variable inside the `llvm-spirv` library set.
Even if `clang-linker-wrapper` knows all the options that need to be passed, we need to come up with how to set all the globals which seems flaky.

The consensuses from other reviewers here is to just use the SPIR-V backend now since it is apparently pretty good, so I'm not planning to use the translator anymore.

AlexeySachkov replied:
You seem to be talking about how to initialize globals within `llvm-spirv/main.cpp` (or whatever  name of the file is), but that shouldn't be necessary. API entry points accept all those options via `TranslatorOpts` object. The translator library does not use any globals to access options

sarnex replied:
Sorry, you're right, I'm too tried. Still, having to parse the strings from the driver and map it to the getter/setter for the object seems not ideal.

AlexeySachkov commented on sycl/doc/design/ThinLTO.md:
```diff
@@ -0,0 +1,149 @@
+framework will run.
+
+It is possible that the number of device images in the fat executable and which
+device image contains which kernel is different with ThinLTO enabled, but we do
+expect this to have any impact on correctness or performance, nor we do expect
```
"but we do expect this..." - missing "not"?

sarnex replied:
yep, thx

AlexeySachkov commented on sycl/doc/design/ThinLTO.md:
```diff
@@ -0,0 +1,149 @@
+fully linked module to accurately compute metadata, see the above explanation of
+SYCL requirement 2). We could get around storing the module by computing the
+metadata inside the LTO framework and storing it for late use by the SYCL
+bundling code, but doing this would require even more SYCL-only customizations
+including even more new function arguments and modifications of the
+`OffloadFile` class. There are also compilations because the LTO framework is
```
Why would it require modifications to `OffloadFile`? I mean, it doesn't seem to have an accessor to `StringData` map stored by `OffloadBinary` contained within `OffloadFile`, but adding such a getter method doesn't seem too SYCL-specific

AlexeySachkov commented on sycl/doc/design/ThinLTO.md:
```diff
@@ -0,0 +1,149 @@
+bundling code, but doing this would require even more SYCL-only customizations
+including even more new function arguments and modifications of the
+`OffloadFile` class. There are also compilations because the LTO framework is
+multithreaded, and not all LLVM data structures are thread safe.
```
It is not clear to me what is there that requires multithreading when each thinLTO thread should be working on its own resulting module alone. Or am I missing something?

AlexeySachkov commented on sycl/doc/design/ThinLTO.md:
```diff
@@ -0,0 +1,149 @@
+The biggest difference here is that we are running CodeGen using the SPIR-V
+backend.
+
+Also, instead of using a lambda function in the `PreCodeGenModuleHook` callback
+to run SYCL finalization passes, we can take advantage of the
+`PreCodeGenPassesHook` field to add passes to the pass manager that the LTO
+framework will run.
```
I got confused by this paragraph, because it contradicts some previous statement. Does it make sense to create sub-sections for short- and long-term designs of thinLTO to better split those two descriptions?

AlexeySachkov commented on sycl/doc/design/ThinLTO.md:
```diff
@@ -0,0 +1,149 @@
+with weak linkage as it cannot be sure that this definition is the correct one.
+Ideally we could remove the weak symbol annotation.
+
+No binary linkage: The SPIR-V target does not currently have a production
+quality binary linker. This means that we must generate a fully linked image as
+part of device linkage. At least for AMD devices, this is not a requirement as
```
SPIR-V may be incomplete, there is nothing in the standard which disallows that. However, producing incomplete SPIR-V  modules in the compiler will require SYCL RT changes so that SYCL RT can pass a set of SPIR-V modules into device JIT compiler which when linked together produce a complete module.

AlexeySachkov reviewed (commented):
Design doc review
